### ISSUE TITLE
Loop Detection Question

### ISSUE THREAD

**Issue by re-fox** [fireeye/capa#191]
I had a question on the feature `characteristic: loop`.

I have a function that does contain a loop but the characteristic appears to not be recognized.

In sample `1d8fd13c890060464019c0f07b928b1a` (Located in the `capa-testfiles` repo) at offset: `0x401920`. 

The function graph contains the following blocks:
![image](https://user-images.githubusercontent.com/57954766/88332303-4d5d3380-ccfc-11ea-922b-d0208344daf5.png)

I wrote a dummy rule to see if I could trigger on the behaviour of a loop
```
rule:
  meta:
    name: loop
    namespace: loop
    scope: function
  features:
    - and:
      - characteristic: loop
```
The only result that I'm getting back is @ `0x402360` and not `0x401920`
```
loop
namespace  loop
scope      function
function @ 0x402360
  and:
    characteristic: loop @ 0x402360
```

Any help or guidance would be appreciated.

**Comment by re-fox** [comment]
If it helps another example of this is in the RC4 section `0x4015F0` of the malware mentioned above `1d8fd13c890060464019c0f07b928b1a`

Removing the `match: contain loop` condition from the `encrypt-data-using-rc4-prga.yml` rule will cause the rule correctly ID the function.  (Shown below)
```
encrypt data using RC4 PRGA
namespace  data-manipulation/encryption/rc4
scope      function
matches    0x4015F0
```

A screenshot of the function, showing a loop: 
![image](https://user-images.githubusercontent.com/57954766/88336205-987a4500-cd02-11ea-9e16-7b37499e06da.png)

**Comment by mike-hunhoff** [comment]
Looks like this is a bug caused by some unexpected behavior in Vivisect.

Vivisect does not set any branch flags for the `jmp` instructions @ `0x401617` and `0x401699`:
```
('0x401617', jmp 0x00401622)
('BR_COND', False)
('BR_FALL', False)
('BR_TABLE', False)
('BR_PROC', False)
('BR_ARCH', False)
('BR_DEREF', False)

('0x40161f', mov dword [ebp - 16],edx)
('BR_COND', False)
('BR_FALL', True)
('BR_TABLE', False)
('BR_PROC', False)
('BR_ARCH', False)
('BR_DEREF', False)

('0x401628', jnc 0x0040169e)
('BR_COND', True)
('BR_FALL', True)
('BR_TABLE', False)
('BR_PROC', False)
('BR_ARCH', False)
('BR_DEREF', False)

('0x401628', jnc 0x0040169e)
('BR_COND', True)
('BR_FALL', False)
('BR_TABLE', False)
('BR_PROC', False)
('BR_ARCH', False)
('BR_DEREF', False)

('0x401699', jmp 0x00401619)
('BR_COND', False)
('BR_FALL', False)
('BR_TABLE', False)
('BR_PROC', False)
('BR_ARCH', False)
('BR_DEREF', False)
```
This causes the following check to ignore the branches @ `0x401617` and `0x401699` causing the loop detection to fail:
 https://github.com/fireeye/capa/blob/583f3f74ecfb6f3b430ba05408b5cb2f2a5c4b3b/capa/features/extractors/viv/function.py#L81

@mr-tz or @williballenthin you guys are more familiar with vivisect. Any thoughts on why Vivisect might not be setting the branch flags for these instructions?

**Comment by williballenthin** [comment]
here are the intended meanings of the branch flags:

```py
# Branch flags (flags returned by the getBranches() method on an opcode)
BR_PROC  = 1<<0  # The branch target is a procedure (call <foo>)
BR_COND  = 1<<1  # The branch target is conditional (jz <foo>)
BR_DEREF = 1<<2  # the branch target is *dereferenced* into PC (call [0x41414141])
BR_TABLE = 1<<3  # The branch target is the base of a pointer array of jmp/call slots
BR_FALL  = 1<<4  # The branch is a "fall through" to the next instruction
BR_ARCH  = 1<<5  # The branch *switches opcode formats*. ( ARCH_FOO in high bits )
```
https://github.com/vivisect/vivisect/blob/402e944e0b1d7ff07ab702c8651a7c1723b24023/envi/__init__.py#L60-L66

given this, i think the behavior of not setting a branch flag is probably correct. i think capa needs to check for the mnemonic `jmp` in the extraction code?

**Comment by mike-hunhoff** [comment]
@williballenthin I think you're correct. I updated my comment after taking another look. Can capa simply check for the branch flags it doesn't want to record specifically `BR_PROC`, `BR_DEREF`, and `BR_ARCH`?

**Comment by mike-hunhoff** [comment]
fixed in #192.
